Kernel chaos nil pointer panic

## Bug Report

Version

```
master
```

Error log in Chaos Mesh controller

```
2020-05-21T19:21:06.192Z	ERROR	controllers.KernelChaos	Get container pid error	{"reconciler": "kernelchaos", "namespace": "chaos-testing", "name": "etcd-577c6ffdd4-wm6xg", "error": "rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 172.16.4.67:31767: connect: no route to host\""}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/pingcap/chaos-mesh/controllers/kernelchaos.(*Reconciler).applyPod
	/src/controllers/kernelchaos/types.go:291
github.com/pingcap/chaos-mesh/controllers/kernelchaos.(*Reconciler).applyAllPods.func1
	/src/controllers/kernelchaos/types.go:259
golang.org/x/sync/errgroup.(*Group).Go.func1
	/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57
2020-05-21T19:21:06.192Z	INFO	util	Creating client to chaos-daemon	{"node": "172.16.4.67"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13a0b3b]

goroutine 444 [running]:
github.com/pingcap/chaos-mesh/controllers/kernelchaos.(*Reconciler).applyPod(0xc0008f0330, 0x1967640, 0xc000169800, 0xc000a71000, 0xc000193400, 0x0, 0x0)
	/src/controllers/kernelchaos/types.go:313 +0x71b
github.com/pingcap/chaos-mesh/controllers/kernelchaos.(*Reconciler).applyAllPods.func1(0xc000534768, 0x0)
	/src/controllers/kernelchaos/types.go:259 +0x4e
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc000796750, 0xc000796840)
	/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:54 +0x66
```

Reproduce steps:

1. create chaos mesh

```
helm3 install chaos-mesh helm/chaos-mesh --namespace=chaos-testing --set bpfki.create=true
```

2. create etcd example

```
kubectl apply -f examples/etcd/etcd.yaml
```

3. create kernel chaos

```
apiVersion: pingcap.com/v1alpha1
kind: KernelChaos
metadata:
  name: kernel-chaos-1
  namespace: chaos-testing
spec:
  duration: 30s
  failKernRequest:
    callchain:
    - funcname: __x64_sys_mount
    failtype: 0
  mode: one
  scheduler:
    cron: '@every 60s'
  selector:
    labelSelectors:
      app: etcd
    namespaces:
    - chaos-testing
  value: ""
```

--- a/controllers/kernelchaos/types.go
+++ b/controllers/kernelchaos/types.go
@@ -188,38 +188,34 @@ func (r *Reconciler) cleanFinalizersAndRecover(ctx context.Context, chaos *v1alp
 func (r *Reconciler) recoverPod(ctx context.Context, pod *v1.Pod, chaos *v1alpha1.KernelChaos) error {
 	r.Log.Info("try to recover pod", "namespace", pod.Namespace, "name", pod.Name)
 
-	c1, err := utils.CreateGrpcConnection(ctx, r.Client, pod, common.Cfg.ChaosDaemonPort)
+	pbClient, err := utils.NewChaosDaemonClient(ctx, r.Client, pod, common.Cfg.ChaosDaemonPort)
 	if err != nil {
 		return err
 	}
-	defer c1.Close()
-
-	daemonClient := pb.NewChaosDaemonClient(c1)
+	defer pbClient.Close()
 
 	if len(pod.Status.ContainerStatuses) == 0 {
 		return fmt.Errorf("%s %s can't get the state of container", pod.Namespace, pod.Name)
 	}
 
-	containerID := pod.Status.ContainerStatuses[0].ContainerID
-
-	containerResponse, err := daemonClient.ContainerGetPid(ctx, &pb.ContainerRequest{
+	containerResponse, err := pbClient.ContainerGetPid(ctx, &pb.ContainerRequest{
 		Action: &pb.ContainerAction{
 			Action: pb.ContainerAction_GETPID,
 		},
-		ContainerId: containerID,
+		ContainerId: pod.Status.ContainerStatuses[0].ContainerID,
 	})
 
 	if err != nil {
 		r.Log.Error(err, "Get container pid error", "namespace", pod.Namespace, "name", pod.Name)
-	} else {
-		r.Log.Info("Get container pid", "namespace", pod.Namespace, "name", pod.Name)
+		return err
 	}
 
-	c2, err := utils.CreateGrpcConnection(ctx, r.Client, pod, common.Cfg.BPFKIPort)
+	r.Log.Info("Get container pid", "namespace", pod.Namespace, "name", pod.Name)
+	conn, err := utils.CreateGrpcConnection(ctx, r.Client, pod, common.Cfg.BPFKIPort)
 	if err != nil {
 		return err
 	}
-	defer c2.Close()
+	defer conn.Close()
 
 	var callchain []*pb_.FailKernRequestFrame
 	for _, frame := range chaos.Spec.FailKernRequest.Callchain {
@@ -230,7 +226,7 @@ func (r *Reconciler) recoverPod(ctx context.Context, pod *v1.Pod, chaos *v1alpha
 		})
 	}
 
-	bpfClient := pb_.NewBPFKIServiceClient(c2)
+	bpfClient := pb_.NewBPFKIServiceClient(conn)
 	_, err = bpfClient.RecoverMMOrBIO(ctx, &pb_.FailKernRequest{
 		Pid:       containerResponse.Pid,
 		Callchain: callchain,
@@ -266,38 +262,33 @@ func (r *Reconciler) applyAllPods(ctx context.Context, pods []v1.Pod, chaos *v1a
 func (r *Reconciler) applyPod(ctx context.Context, pod *v1.Pod, chaos *v1alpha1.KernelChaos) error {
 	r.Log.Info("Try to inject kernel on pod", "namespace", pod.Namespace, "name", pod.Name)
 
-	c1, err := utils.CreateGrpcConnection(ctx, r.Client, pod, common.Cfg.ChaosDaemonPort)
+	pbClient, err := utils.NewChaosDaemonClient(ctx, r.Client, pod, common.Cfg.ChaosDaemonPort)
 	if err != nil {
 		return err
 	}
-	defer c1.Close()
-
-	pbClient := pb.NewChaosDaemonClient(c1)
+	defer pbClient.Close()
 
 	if len(pod.Status.ContainerStatuses) == 0 {
 		return fmt.Errorf("%s %s can't get the state of container", pod.Namespace, pod.Name)
 	}
 
-	containerID := pod.Status.ContainerStatuses[0].ContainerID
-
 	containerResponse, err := pbClient.ContainerGetPid(ctx, &pb.ContainerRequest{
 		Action: &pb.ContainerAction{
 			Action: pb.ContainerAction_GETPID,
 		},
-		ContainerId: containerID,
+		ContainerId: pod.Status.ContainerStatuses[0].ContainerID,
 	})
-
 	if err != nil {
 		r.Log.Error(err, "Get container pid error", "namespace", pod.Namespace, "name", pod.Name)
-	} else {
-		r.Log.Info("Get container pid", "namespace", pod.Namespace, "name", pod.Name)
+		return err
 	}
 
-	c2, err := utils.CreateGrpcConnection(ctx, r.Client, pod, common.Cfg.BPFKIPort)
+	r.Log.Info("Get container pid", "namespace", pod.Namespace, "name", pod.Name)
+	conn, err := utils.CreateGrpcConnection(ctx, r.Client, pod, common.Cfg.BPFKIPort)
 	if err != nil {
 		return err
 	}
-	defer c2.Close()
+	defer conn.Close()
 
 	var callchain []*pb_.FailKernRequestFrame
 	for _, frame := range chaos.Spec.FailKernRequest.Callchain {
@@ -308,7 +299,7 @@ func (r *Reconciler) applyPod(ctx context.Context, pod *v1.Pod, chaos *v1alpha1.
 		})
 	}
 
-	bpfClient := pb_.NewBPFKIServiceClient(c2)
+	bpfClient := pb_.NewBPFKIServiceClient(conn)
 	_, err = bpfClient.FailMMOrBIO(ctx, &pb_.FailKernRequest{
 		Pid:         containerResponse.Pid,
 		Ftype:       pb_.FailKernRequest_FAILTYPE(chaos.Spec.FailKernRequest.FailType),

--- a/pkg/utils/chaosdaemon.go
+++ b/pkg/utils/chaosdaemon.go
@@ -1,3 +1,16 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package utils
 
 import (

--- a/pkg/utils/chaosdaemon_test.go
+++ b/pkg/utils/chaosdaemon_test.go
@@ -1,3 +1,16 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package utils
 
 import (

